sklearn/datasets/_arff_parser.py

```diff
@@ -0,0 +1,501 @@
+
+    output_array_type : {"numpy", "sparse", "pandas"}
+        The type of the arrays that will be returned. The possibilities ara:
+
+        - `"numpy"`: both `X` and `y` will be NumPy arrays;
+        - `"sparse"`: `X` will be sparse matrix and `y` will be a NumPy array;
+        - `"pandas"`: `X` will be a pandas DataFrame and `y` will be either a
+          pandas Series or DataFrame.
+
+    columns_info : dict
+        The information provided by OpenML regarding the columns of the ARFF
+        file.
+
+    feature_names_to_select : list of str
+        A list of the feature names to be selected.
+
+    target_names_to_select : list of str
+        A list of the target names to be selected.
+
+    infer_casting : bool
+        Whether or not to infer the type of the data in each column and thus
+        cast the data to the most appropriate type. Activating this option
+        will be more costly but will solved ambiguities regarding the data
+        types provided by the ARFF metadata.
+
+    Returns
+    -------
+    X : {ndarray, sparse matrix, dataframe}
+        The data matrix.
+
+    y : {ndarray, dataframe, series}
+        The target.
+
+    frame : dataframe or None
+        A dataframe containing both `X` and `y`. `None` if
+        `output_array_type != "pandas"`.
+
+    nominal_attributes : list of str or None
+        The names of the features that are categorical. `None` if
+        `output_array_type == "pandas"`.
+    """
+
+    def _io_to_generator(gzip_file):
+        for line in gzip_file:
+            yield line.decode("utf-8")
+
+    stream = _io_to_generator(gzip_file)
+
+    return_type = _arff.COO if output_arrays_type == "sparse" else _arff.DENSE_GEN
+    arff_container = _arff.load(stream, return_type=return_type)
+
+    nominal_attributes = _compute_nominal_attributes(
+        columns_info_openml, feature_names_to_select, target_names_to_select
+    )
+    columns_to_select = feature_names_to_select + target_names_to_select
+    if output_arrays_type == "pandas":
+        pd = check_pandas_support("fetch_openml with as_frame=True")
+
+        columns_info = OrderedDict(arff_container["attributes"])
+        columns_names = list(columns_info.keys())
+
+        # calculate chunksize
+        first_row = next(arff_container["data"])
+        first_df = pd.DataFrame([first_row], columns=columns_names)
+
+        row_bytes = first_df.memory_usage(deep=True).sum()
+        chunksize = get_chunk_n_rows(row_bytes)
+
+        # read arff data with chunks
+        columns_to_keep = [col for col in columns_names if col in columns_to_select]
+        dfs = [first_df[columns_to_keep]]
+        for data in _chunk_generator(arff_container["data"], chunksize):
+            dfs.append(pd.DataFrame(data, columns=columns_names)[columns_to_keep])
+        frame = pd.concat(dfs, ignore_index=True)
+        del dfs, first_df
+
+        frame = _cast_frame(frame, columns_info_openml, infer_casting)
```

Comment:
This is the only difference with the previous implementation where we attempt to infer casting if requested.